File: 02_DrugiPrimer/main.c
/*
 * Deljeno stanje:
 * Promenljiva counter koju sve niti uvecavaju.
 *
 * Tip race condition-a:
 * Vise niti istovremeno citaju i upisuju istu promenljivu
 * bez sinhronizacije, pa se neka uvecavanja izgube.
 *
 * Ocekivano ponasanje:
 * Program pokrece 2 niti, svaka uvecava counter 1000000 puta.
 * Ocekujemo da counter na kraju bude 2000000.
 * Sa mutexom, rezultat je uvek tacan.
 * Bez muteksa, rezultat bi bio manji od 2000000.
 */

#include <stdio.h>
#include <pthread.h>

int counter = 0;
pthread_mutex_t mutex;

void *posao(void *args) {
    for (int i = 0; i < 1000000; i++) {
        pthread_mutex_lock(&mutex);   /* zakljucaj - ulaz u kriticnu sekciju */
        counter++;                     /* kriticna sekcija */
        pthread_mutex_unlock(&mutex); /* otkljucaj - izlaz iz kriticne sekcije */
    }
    return NULL;
}

int main() {
    pthread_mutex_init(&mutex, NULL);

    pthread_t t1, t2;
    pthread_create(&t1, NULL, posao, NULL);
    pthread_create(&t2, NULL, posao, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

    printf("counter = %d\n", counter);

    pthread_mutex_destroy(&mutex);
    return 0;
}

/*
 * Zasto moze doci do race condition-a:
 * counter++ se izvrsava u tri koraka: ucitaj, uvecaj, upisi.
 * Ako dve niti istovremeno ucitaju istu vrednost, jedno
 * uvecanje se izgubi jer obe upisuju isti rezultat.
 *
 * Kriticna sekcija:
 * Linija counter++ jer tu dolazi do pristupa deljenoj promenljivoj.
 *
 * Kako je izvrsena sinhronizacija:
 * Koristimo pthread_mutex_t. Pre counter++ nit zakljuca mutex,
 * a posle otkljuca. Dok jedna nit drzi mutex, druga ceka.
 * Tako u svakom trenutku samo jedna nit izvrsava counter++.
 *
 * Promena ponasanja nakon sinhronizacije:
 * Bez mutexa: rezultat je razlicit pri svakom pokretanju i manji od 2000000.
 * Sa mutexom: rezultat je uvek tacno 2000000.
 * Program je sporiji jer niti cekaju jedna drugu, ali je korektan.
 */

File: 02_DrugiPrimer/main.go
/*
 * Deljeno stanje:
 * Promenljiva counter koju sve goroutine uvecavaju.
 *
 * Tip race condition-a:
 * Vise goroutina istovremeno citaju i upisuju istu promenljivu
 * bez sinhronizacije, pa se neka uvecavanja izgube.
 *
 * Ocekivano ponasanje:
 * Program pokrece 2 goroutine, svaka uvecava counter 1000000 puta.
 * Ocekujemo da counter na kraju bude 2000000.
 * Sa mutexom, rezultat je uvek tacan.
 *
 * -----------------------------------------------------------
 * C vs Go:
 *   pthread_mutex_t mutex               ->  var mu sync.Mutex
 *   pthread_mutex_init(&mutex, NULL)    ->  (automatski inicijalizovan)
 *   pthread_mutex_lock(&mutex)          ->  mu.Lock()
 *   pthread_mutex_unlock(&mutex)        ->  mu.Unlock()
 *   pthread_mutex_destroy(&mutex)       ->  (automatski, GC)
 * -----------------------------------------------------------
 */

package main

import (
	"fmt"
	"sync"
)

var counter int
var mu sync.Mutex

func posao(wg *sync.WaitGroup) {
	defer wg.Done()
	for i := 0; i < 1000000; i++ {
		mu.Lock()   // zakljucaj - ulaz u kriticnu sekciju
		counter++   // kriticna sekcija
		mu.Unlock() // otkljucaj - izlaz iz kriticne sekcije
	}
}

func main() {
	var wg sync.WaitGroup

	wg.Add(2)
	go posao(&wg)
	go posao(&wg)

	wg.Wait()

	fmt.Printf("counter = %d\n", counter)
}

/*
 * Zasto moze doci do race condition-a:
 * counter++ se izvrsava u tri koraka: ucitaj, uvecaj, upisi.
 * Ako dve goroutine istovremeno ucitaju istu vrednost, jedno
 * uvecanje se izgubi jer obe upisuju isti rezultat.
 *
 * Kriticna sekcija:
 * Linija counter++ jer tu dolazi do pristupa deljenoj promenljivoj.
 *
 * Kako je izvrsena sinhronizacija:
 * Koristimo sync.Mutex. Pre counter++ goroutina zakljuca mutex,
 * a posle otkljuca. Dok jedna goroutina drzi mutex, druga ceka.
 *
 * Promena ponasanja nakon sinhronizacije:
 * Bez mutexa: rezultat je razlicit pri svakom pokretanju.
 * Sa mutexom: rezultat je uvek tacno 2000000.
 */
